File: python/01010_pairs_of_songs_with_total_durations_divisible_by_60.py
from typing import List


#       Python
# Time complexity - O (N)
# Space complexity - O (1)
class Solution:
    def numPairsDivisibleBy60 (self, time: List[int]) -> int:
        remainder_count = {}
        result = 0

        for t in time:
            remainder = t % 60
            complement = (60 - remainder) % 60
            result += remainder_count.get(complement, 0)
            remainder_count[remainder] = remainder_count.get(remainder, 0) + 1
        return result

File: java/01010_pairs_of_songs_with_total_durations_divisible_by_60.java
import java.util.HashMap;
import java.util.Map;

// Time complexity - O (N)
// Space complexity - O (1)
public class PairsOfSongsWithTotalDurationsDivisibleBy60 {

    public static int numPairsDivisibleBy60(int[] time) {
        Map<Integer, Integer> remainderCounts = new HashMap<>();
        int result = 0;

        for (int t : time) {
            int remainder = t % 60;
            int complement = (60 - remainder) % 60;
            result += remainderCounts.getOrDefault(complement, 0);
            remainderCounts.put(remainder, remainderCounts.getOrDefault(remainder, 0) + 1);
        }

        return result;
    }
}
